### PR TITLE
Assignment2

### PR DESCRIPTION
**Pull request by CDobbins1**

### CODE REVIEW

**Comment by JerettLatimer** [comment]
Will review!

File: SecretSanta/src/SecretSanta.Data/FingerPrintEntityBase.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SecretSanta.Data
+{
+    public class FingerPrintEntityBase : EntityBase
+    {
+        private string _CreatedBy = string.Empty;
```
Comment (JerettLatimer) [inline SecretSanta/src/SecretSanta.Data/FingerPrintEntityBase.cs]: Good job null checking here.

File: SecretSanta/test/SecretSanta.Data.Tests/UserGroupTests.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SecretSanta.Data.Tests
+{
+    [TestClass]
+    public class UserGroupTests : TestBase
+    {
+        [TestMethod]
+        public async Task CreateUser_WithManyGroups()
```
Comment (JerettLatimer) [inline SecretSanta/test/SecretSanta.Data.Tests/UserGroupTests.cs]: Not a huge deal, but I did notice that your unit tests in this class have a different naming convention compared to your other test classes, which follow the Microsoft standards.

File: SecretSanta/test/SecretSanta.Data.Tests/UserGroupTests.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            using (ApplicationDbContext dbContext = new ApplicationDbContext(Options, httpContextAccessor))
+            {
+                var retrievedUser = await dbContext.Users.Where(u => u.Id == user.Id)
+                    .Include(u => u.UserGroups).ThenInclude(ug => ug.Group).SingleOrDefaultAsync();
+
+                Assert.IsNotNull(retrievedUser);
+                Assert.AreEqual<int>(2, retrievedUser.UserGroups.Count);
+                Assert.IsNotNull(retrievedUser.UserGroups[0].Group);
+                Assert.IsNotNull(retrievedUser.UserGroups[1].Group);
+                Assert.AreEqual<string>("TestGroup1", retrievedUser.UserGroups[0].Group.Name);
```
Comment (JerettLatimer) [inline SecretSanta/test/SecretSanta.Data.Tests/UserGroupTests.cs]: Good job using explicit Assert statements.

**Comment by breaman** [comment]
##  (CDobbins1)

## Assignment2

| Project Analysis Summary |   |
| -----------------| - |
| Nullable Enabled: | **✔** |
| Unexpected Disabled Warnings:  | **✔**
| Appropriately handle all warnings: | **✔** |

| ProjectName | Nullable | Solutions | Analyzers | Unexpected Disabled Warnings | Waning Level |
| -- | -- | -- | -- | -- | -- |
| SecretSanta.Api | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Business | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Data | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Web | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer |  | 4 |
| SecretSanta.Api.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Business.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Data.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer, Microsoft.EntityFrameworkCore.Analyzers |  | 4 |
| SecretSanta.Web.Tests | Enable | SecretSanta.sln | IntelliTect Analyzers, Microsoft CodeQuality Analyzers, Microsoft .NET Core Analyzers, Microsoft .NET Framework Analyzers, Microsoft.CodeAnalysis.VersionCheckAnalyzer |  | 4 |


## Main Assignment

1. Add the following classes to the **SecretSanta.Data** project with the following associated properties:
   - **`EntityBase`** ✔
     - `int Id`
   - **`FingerPrintEntityBase`** inherit from **`EntityBase`** ✔
     - `string CreatedBy`
     - `DateTime CreatedOn`
     - `string ModifiedBy`
     - `DateTime ModifiedOn`
   - **`Group`** inherit from **`FingerPrintEntityBase`** ✔
     - `string Name`
   - Create a class that will hold the many to many relationship between User and Group ✔
     - Add a collection on both User and Group that references this class
2. Modify the following classes in the **SecretSanta.Data** project:
   - **`Gift`** ✔
     - Remove the Id property and inherit from FingerPrintEntityBase
   - **`User`** ✔
     - Remove the Id property and inherit from FingerPrintEntityBase
     - Make `Gifts` read/write
     - Create a nullable Santa property that contains a User as someone's Santa
3. Remove non-default constructors for `Gift` and `User`. ✔
4. In the **SecretSanta.Data** project
   - Create the **`ApplicationDbContext`** class ✔
     - Create two constructors ✔
        - One only takes the DbContextOptions and passes it on to the base class
        - One that takes the DbContextOptions and an IHttpContextAccessor, pass the DbContextOptions to the base class and save off the IHttpContextAccessor in a class property
     - Add DbSet properties for User, Gift, and Group classes ✔
     - Override OnModelCreating method and use the fluent API to configure the many-to-many pieces that are needed for setting up the database properly ✔
     - Override both SaveChanges methods and apply the fingerprinting logic ✔
5. Update unit tests in the **`SecretSanta.Data.Tests`** project to test actual database logic instead of just constructor logic
   - Mock up IHttpContextAccessor to verify that fingerprint information is working correctly (using Moq library) ✔


## Extra Credit

The following are options for extra credit:
- Enable console logging that will log all database commands ✔

**Pull request closed without merging**